### PR TITLE
Evaluate opt-in TreeDB HNSW wavefront batching

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Evaluate an internal-only, opt-in "wavefront" layer-0 traversal mode for TreeDB `column_graph` native search, measure recall/performance tradeoffs across widths 2/4/8/16, and make a data-driven decision on public exposure.

## Context

Issue #1981 called for a prototype wavefront batching traversal and benchmark sweep. This PR delivers the internal prototype, sweep evidence, and a closeout decision for the #1973 workstream.

Milestones covered:
- M0: reproduced baseline/profile on this branch and recorded artifacts.
- M1: added fixed-query-set result/recall delta harness before traversal prototype commits.
- M2/M3: prototyped internal, non-default wavefront traversal and swept widths 2/4/8/16.
- M4 decision: **defer public exposure**. The prototype remains internal-only for benchmark evidence; no public API is added.

## Latest head / review fix update (2026-06-02)

- Latest head after all review fixes: `a396a17ca72299bb4690bf4bc57ecff0b50db42c`.
- Review fixes landed:
  - `c06fd86663ecbcc21cafe2aefb139c89ad88a1bf` removed the unused `wavefrontEnabled()` helper.
  - `e6ea6085ab9d0e3b176fc57ae943b96de37f776f` fails closed when `WavefrontWidth > normalized ef_search` and adds a regression assertion.
  - `e6ea6085ab9d0e3b176fc57ae943b96de37f776f` removes redundant `_ = candidateOrdinal` test noise.
  - `7203fad25e710ce0313680b658ca673fb5ae5002` removes redundant `ScoreOverlapCount` field/increment from the result-delta harness.
  - `a396a17ca72299bb4690bf4bc57ecff0b50db42c` expands a non-empty partial wave before fallback seeding and adds `TestColumnVectorGraphWavefrontProcessesPartialWaveBeforeSeeding1981`.
- Latest local artifact directory: `/tmp/treedb_wavefront_1981_reviewfix2_20260602_124156`.
- Latest local validation reran `GOWORK=off go test ./TreeDB/collections -count=1`, focused wavefront tests, the core/public stats-mode benchmarks, the #1979 batchability benchmark, and the wavefront sweep.
- Exact/default latest-head evidence remains healthy: core full diagnostics `12479 ns/op`, core minimal `12114 ns/op`, public production `13103 ns/op`, batchability `6393 ns/op`; exact/default paths keep wavefront counters at 0 and no graph-row fallback.
- Latest minimal-stats wavefront sweep: exact `7163 ns/op`/100.00% recall; width=2 `6818 ns/op`/98.33%; width=4 `6569 ns/op`/91.67%; width=8 `6767 ns/op`/85.00%; width=16 `7465 ns/op`/83.33%; all `0 B/op`, `0 allocs/op`.
- Performance assessment remains: no exact/default regression; wavefront is internal-only and public exposure stays deferred because recall/result drift is measurable.
- Latest-head GitHub Actions CI is green for `a396a17ca72299bb4690bf4bc57ecff0b50db42c`. One unrelated TreeDB `race-check (linux)-1` job initially failed in `TreeDB/mongo_gateway` and passed on failed-job rerun; all current check rollup entries are passing.

## Non-goals

- No public API exposure of the wavefront traversal mode.
- No storage or durable metadata changes.
- No default behavior change.
- No dot-product kernel work.

## Design

- Adds `columnVectorGraphNativeSearchTraversalModeWavefront` as an internal constant for prepared typed-column `column_graph` search only.
- Default/exact traversal is completely unchanged; wavefront is explicit opt-in and fail-closed for invalid mode/width and non-prepared readers.
- `columnVectorGraphNativeSearchTraversalOptions` validates mode and width and is the sole resolution path for traversal mode (no helper aliases).
- Wavefront counters (rounds, staged neighbors, max tile size) extend the existing stats/bench reporting without touching the exact path.
- Result-delta harness (`columnVectorGraphResultDelta1981`) tracks `TopKOverlap`, `OrderingMismatches`, `MaxScoreAbsDelta`, `ExactOnly`, `CandidateOnly`, and `CandidateResultShorter`; no redundant fields.

## Scope

- Includes: `TreeDB/collections/column_vector_graph_search.go`, wavefront test and benchmark files in `TreeDB/collections`.
- Excludes: public API surface, storage format, durable metadata, upper-layer traversal.

## Correctness

- Tests run:

```sh
GOWORK=off go test ./TreeDB/collections -count=1

GOWORK=off go test ./TreeDB/collections \
  -run 'TestColumnVectorGraph(WavefrontOptions1981|WavefrontRequiresPrepared1981|ResultDeltaHarness1981)' \
  -count=1

GOWORK=off go test ./TreeDB/collections \
  -run 'TestColumnVectorGraphNativeSearchEmptyAndTopKClampV3|TestColumnVectorGraphNativeSearchRequiresScratchV3' \
  -count=1
```

All passed locally.

- Corruption/edge cases covered: fail-closed on invalid mode, fail-closed on invalid width, fail-closed for non-prepared readers.
- Concurrency/race coverage: N/A (no new concurrency paths).

## Performance

- Benchmarks run:

```sh
GOWORK=off go test ./TreeDB/collections \
  -run '^$' \
  -bench '^BenchmarkColumnVectorGraphNativeSearchCosineTypedColumnStatsMode2042/(full_diagnostics|minimal)$' \
  -benchtime=2s -count=1 -benchmem

GOWORK=off go test ./TreeDB/collections \
  -run '^$' \
  -bench '^BenchmarkOpenVectorIndexSearcherColumnGraphTypedColumnNativeReaderV4StatsMode2126/(stats=full_diagnostics|stats=production)$' \
  -benchtime=2s -count=1 -benchmem

GOWORK=off go test ./TreeDB/collections \
  -run '^$' \
  -bench '^BenchmarkColumnVectorGraphSearchBatchability1979/ef_search_128/score=indexed$' \
  -benchtime=2s -count=1 -benchmem

GOWORK=off go test ./TreeDB/collections \
  -run '^$' \
  -bench '^(BenchmarkColumnVectorGraphWavefrontSearch1981|BenchmarkColumnVectorGraphWavefrontSearchMinimal1981)' \
  -benchtime=2s -count=1 -benchmem
```

- Exact/default regression assessment (single-run local; Apple M3, darwin/arm64, Go go1.25.5):

| Boundary | Benchmark | Base ns/op | Candidate ns/op | B/op | allocs/op | Assessment |
| --- | --- | ---:| ---:| ---:| ---:| --- |
| Core graph-only full diagnostics | `BenchmarkColumnVectorGraphNativeSearchCosineTypedColumnStatsMode2042/full_diagnostics` | 12395 | 12351 | 0 | 0 | no material regression |
| Core graph-only minimal | `.../minimal` | 12369 | 12083 | 0 | 0 | no material regression |
| Public typed-column production stats | `BenchmarkOpenVectorIndexSearcherColumnGraphTypedColumnNativeReaderV4StatsMode2126/stats=production` | 13402 | 12970 | 784 | 2 | no material regression |
| Batchability fixture | `BenchmarkColumnVectorGraphSearchBatchability1979/ef_search_128/score=indexed` | 7995 | 6692 | 0 | 0 | counters unchanged |

Source-health stayed healthy in exact/default paths: prepared graph views = 1, vector/norm/adjacency mmap/direct, graph-row fallbacks = 0, wavefront counters = 0.

- Wavefront sweep (fixture: 8192 rows, 128 dims, degree 16, topK 10, efSearch 128, 6 fixed query ordinals; exact mode is oracle):

| Mode | ns/op | score batch calls/search | recall vs exact | ordering mismatches/query |
| --- | ---:| ---:| ---:| ---:|
| exact | 6987 | 29.17 | 100.00% | 0.000 |
| wavefront width=2 | 6857 | 23.83 | 98.33% | 0.333 |
| wavefront width=4 | 6775 | 18.17 | 95.00% | 1.167 |
| wavefront width=8 | 8207 | 17.00 | 81.67% | 4.000 |
| wavefront width=16 | 11075 | 23.00 | 81.67% | 3.333 |

- Regressions: none on exact/default path. Wavefront width ≥ 8 shows throughput regression and significant recall drop; width 2/4 show marginal speedup with measurable recall drift.

## Rollout / Toggle Plan

- Default setting: exact traversal (unchanged). Wavefront is explicit opt-in via internal `columnVectorGraphNativeSearchTraversalModeWavefront` constant; not reachable from any public API.
- How to disable quickly: N/A — wavefront is not on by default and has no public knob.

## Follow-ups

- Public exposure of wavefront deferred pending further recall/robustness improvements (#1973 closeout decision).

### Copied/Adapted From Old Stack

- Copied: N/A
- Adapted: N/A
- Oracle/comparator only: fixed-query result-delta harness used as oracle comparator
- Quarantined/not copied: N/A

### Path Identity

- Native reader: prepared typed-column path only
- Decoded comparator: N/A
- Legacy/native vector index: N/A
- Unsupported/fail-closed: wavefront fails closed for non-prepared readers, invalid mode, invalid width

### Base And Dependency State

- Base branch: main
- Base commit: `7e270ed61cfdad50c15f652e245604e1afc9852f`
- Base PR/head: N/A
- #1621/#1634 assumptions: N/A
- Missing generic column-store primitives: N/A
- Parent review fixes propagated: N/A

### Evidence

- Tests: see Correctness section above; all passed locally
- Benchmarks: see Performance section above
- Status/fallback proof: exact/default path counters unchanged; wavefront counters = 0 on default runs
- Performance attribution: wavefront width 2/4 narrow speedup in minimal-stats mode; width ≥ 8 regresses throughput and recall
- Local regressions found/fixed: none

### Test Plan Start

- Milestone tasks targeted: M1 (result-delta harness), M2/M3 (wavefront prototype + sweep)
- Tests to add before or with implementation: fixed-query result-delta harness, wavefront options validation, requires-prepared guard
- Existing tests to preserve: all existing `column_vector_graph_search` tests
- #1646 test-list changes made before implementation: N/A

### Performance Plan Start

- Relevant benchmarks/metrics for this PR: wavefront sweep, exact/default regression boundary, batchability fixture
- Baseline/comparator command: see Performance section
- Expected setup/search/materialization boundary: search layer 0 only
- Upstream costs explicitly out of scope: upper-layer traversal, materialization
- Local performance risks to watch: recall drift at width ≥ 4; throughput regression at width ≥ 8

### Test Plan Close

- Additional tests found during implementation/review: N/A
- Tests added or changed: `TestColumnVectorGraphWavefrontOptions1981`, `TestColumnVectorGraphWavefrontRequiresPrepared1981`, `TestColumnVectorGraphResultDeltaHarness1981`
- Failing tests fixed: N/A
- #1646 test-list changes made at close: N/A
- Residual untested gaps: none

### Performance Plan Close

- Benchmark commands run: see Performance section
- Results summary: width 2/4 show 1–3% ns/op improvement with 2–5% recall drop; width ≥ 8 regress on both axes
- Before/after or baseline comparison: see Performance section tables
- Local slow/heavy code found and fixed: N/A
- Remaining upstream or deferred costs: public wavefront exposure deferred

### AI Review Loop

- Codex review: latest-head refresh after `a396a17ca72299bb4690bf4bc57ecff0b50db42c` reported no major issues. Earlier Codex partial-wave fallback finding was fixed by `a396a17ca72299bb4690bf4bc57ecff0b50db42c` with a regression test.
- Copilot review: latest-head review for `a396a17ca72299bb4690bf4bc57ecff0b50db42c` generated no new comments. Earlier findings were fixed (`wavefrontEnabled()` removed, wavefront width bounded after normalized `ef_search`, redundant `_ = candidateOrdinal` removed, redundant `ScoreOverlapCount` removed).
- CodeRabbit review: status context reports pass/skipped, but the substantive PR comment available locally is a rate/credits-limit warning; no CodeRabbit findings to act on.
- Review comments/threads resolved or explicitly dismissed: all Codex/Copilot threads replied to and resolved.
- Re-requested after final meaningful push: Codex refreshed after `a396a17ca72299bb4690bf4bc57ecff0b50db42c`; Copilot also produced a latest-head no-new-comments review.

---

### Scope
- [x] Single, clear objective for this PR (not a bundle)
- [x] Explicit include list (files/symbols touched): `column_vector_graph_search.go`, wavefront test/bench files in `TreeDB/collections`
- [x] Explicit exclude list (files/symbols NOT touched): public API, storage format, durable metadata, upper-layer traversal
- [x] No unwanted feature reintroduction unless explicitly stated in the PR objective

### Safety / Correctness
- [x] Clear written-vs-durable semantics for any `*Sync` path touched: N/A
- [x] No new mmap usage on mutable/truncating files
- [x] All new length fields are capped before allocation (OOM-safe): N/A
- [x] CRC/checksum behavior is fail-closed (clean error, no panic): N/A
- [x] Any concurrency change has a defined state machine and invariants: N/A

### Tests
- [x] Added deterministic regression tests for the targeted failure mode
- [x] Tests avoid sleeps where possible (use latches/hooks)
- [x] Added corruption/edge-case tests (invalid mode, invalid width, non-prepared reader fail-closed)
- [x] Ran locally:
  - [x] `GOWORK=off go test ./TreeDB/collections -count=1`
  - [x] Targeted packages/benchmarks: see Correctness and Performance sections
- [x] Broader latest-head GitHub Actions CI is green for `a396a17ca72299bb4690bf4bc57ecff0b50db42c`

### Benchmarks / Measurements
- [x] Added or updated a benchmark relevant to the change
- [x] Reported results in PR description (before/after, command, machine)
- [ ] Included "incompressible" (worst-case) results if compression is involved: N/A

### Docs
- [ ] Updated `docs/OPT_SPRINT_NEXT.md` milestone status (if applicable)
- [x] Documented any new config knobs + defaults: wavefront is internal-only, no public knob
- [x] Called out any format change in PR description: no format changes

### Rollout / Toggle Plan (if behavior-affecting)
- [x] Safe defaults (feature off or conservative threshold): exact traversal is default
- [x] Clear knobs to disable/revert behavior quickly: N/A (not on by default)
- [x] Observability: wavefront counters in stats reporting

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

File: TreeDB/collections/column_vector_graph_search.go [hunk 1]
```diff
@@ -76,6 +88,60 @@ func columnVectorGraphScoreBatchModeForSearchPlan(requested columnVectorGraphSco
...
+		return columnVectorGraphNativeSearchTraversalModeWavefront, nil
+	default:
+		return columnVectorGraphNativeSearchTraversalModeExact, errColumnVectorGraphNativeSearchTraversalModeInvalid
+	}
+}
+
+func (m columnVectorGraphNativeSearchTraversalMode) wavefrontEnabled() bool {
+	mode, err := m.normalized()
+	return err == nil && mode == columnVectorGraphNativeSearchTraversalModeWavefront
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_search.go]: Fixed by `c06fd86663ecbcc21cafe2aefb139c89ad88a1bf`: removed the unused `wavefrontEnabled()` helper so traversal mode resolution stays centralized in `columnVectorGraphNativeSearchTraversalOptions`/`normalized()`.

File: TreeDB/collections/column_vector_graph_search.go [hunk 2]
```diff
@@ -963,7 +1052,8 @@ func (r *columnVectorGraphPhysicalRowReader) SearchCosine(query []float32, opts
 	if degree < 0 {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_search.go:1050]: Fixed by `e6ea6085ab9d0e3b176fc57ae943b96de37f776f`: wavefront now fails closed when `wavefrontWidth > efSearch` after `efSearch` is normalized/clamped, with a regression assertion in `TestColumnVectorGraphWavefrontOptions1981`.

File: TreeDB/collections/column_vector_graph_wavefront_1981_test.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+			dst.OrderingMismatches++
+		}
+		if candidateScore, ok := candidateScores[exactOrdinal]; ok {
+			dst.TopKOverlap++
+			dst.ScoreOverlapCount++
+			if delta := math.Abs(candidateScore - exact[i].Score); delta > dst.MaxScoreAbsDelta {
+				dst.MaxScoreAbsDelta = delta
+			}
+		}
+		_ = candidateOrdinal
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_wavefront_1981_test.go]: Fixed by `e6ea6085ab9d0e3b176fc57ae943b96de37f776f`: removed the redundant `_ = candidateOrdinal` line.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

File: TreeDB/collections/column_vector_graph_search.go [hunk 4]
```diff
@@ -1196,6 +1298,126 @@ func (r *columnVectorGraphPhysicalRowReader) SearchCosine(query []float32, opts
...
+		wave := scratch.wavefrontCandidates[:0]
+		for len(wave) < wavefrontWidth && *visitedCandidates < candidateLimit {
+			var candidate columnVectorGraphSearchCandidate
+			var ok bool
+			if debugCounters != nil {
+				candidate, ok = scratch.popFrontierDebug(debugCounters)
+			} else {
+				candidate, ok = scratch.popFrontier()
+			}
+			if !ok {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_search.go:1318]: Fixed by `a396a17ca72299bb4690bf4bc57ecff0b50db42c`: when `popFrontier` misses after staging a non-empty wave, the loop now breaks and expands the staged candidates before fallback seeding. Added `TestColumnVectorGraphWavefrontProcessesPartialWaveBeforeSeeding1981` to cover the sparse/partial-wave budget case.

File: TreeDB/collections/column_vector_graph_wavefront_1981_test.go [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+	"testing"
+)
+
+type columnVectorGraphResultDelta1981 struct {
+	Queries                int
+	TopK                   int
+	TopKOverlap            int
+	OrderingMismatches     int
+	ScoreOverlapCount      int
+	MaxScoreAbsDelta       float64
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_wavefront_1981_test.go:14]: Fixed by `7203fad25e710ce0313680b658ca673fb5ae5002`: removed the redundant `ScoreOverlapCount` field and increment from the result-delta harness.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
Final mergeable-candidate status for #1981 at `a396a17ca72299bb4690bf4bc57ecff0b50db42c`:

- Local validation artifacts: `/tmp/treedb_wavefront_1981_reviewfix2_20260602_124156`.
- Local validation passed:
  - `GOWORK=off go test ./TreeDB/collections -count=1`
  - focused wavefront tests including `TestColumnVectorGraphWavefrontProcessesPartialWaveBeforeSeeding1981`
  - core/public stats-mode benchmarks, #1979 batchability benchmark, and wavefront sweep.
- Latest-head CI: green. Initial `TreeDB: vet+test / race-check (linux)-1` failed in unrelated `TreeDB/mongo_gateway` `command_wal_relaxed`; failed-job rerun passed and current check rollup is all passing.
- AI review: Codex latest-head review reports no major issues; Copilot latest-head review generated no new comments; CodeRabbit status is pass/skipped but substantive comment remains rate/credits-limited.
- Review threads: all Codex/Copilot threads replied to and resolved.
- Performance assessment: exact/default paths show no material regression; wavefront remains internal-only and public exposure is deferred due measurable recall/result drift.

Not merging directly per coordinator instruction.
